tests/fixtures/ddp_fixtures.py

```diff
@@ -1,70 +1,55 @@
-            if k not in original_environ:
-                keys_to_remove.append(k)
-        for k in keys_to_remove:
-            del os.environ[k]
-
-        # reset the other keys
-        for k, v in original_environ.items():
-            os.environ[k] = v
-
-
-@pytest.fixture(autouse=True)
-def ddp_fork_multi_pytest(request: _pytest.fixtures.FixtureRequest, reset_osenviron: None) -> None:
-    # When using DDP fork, the same command is executed again.
-    # When multiple tests are executed with a single pytest command,
-    # then the DDP fork will re-run all tests, when instead we just want
-    # the forked test to run again. To prevent this behavior, we store the
-    # running pytest id as an environ. If this variable is set, and it does
-    # NOT match the current pytest test ID, then the test should be skipped
-    # as the test was launched by a different DDP fork
-    test_name: str = request.node.name
-    if "PYTEST_TEST_ID" not in os.environ:
-        os.environ["PYTEST_TEST_ID"] = test_name
-        return
-    if os.environ["PYTEST_TEST_ID"] != test_name:
-        pytest.skip("Skipping test invoked within a DDP fork")
-
-
-@pytest.fixture
-def ddp_tmpdir(tmpdir: pathlib.Path) -> str:
-    if os.environ.get("DDP_TMP_DIR") is None:
-        # Need to set the tmpdir as an environ, so all subprocesses share the same tmpdir.
-        # Upon subprocess calls, pytest gives each ddp process a different tmpdir
-        # Instead, we want to use the same tmpdir for test case assertions.
-        os.environ['DDP_TMP_DIR'] = str(tmpdir)
-        return str(tmpdir)
-    return os.environ["DDP_TMP_DIR"]
+def ddp_cleanup():
+    yield
+    if torch.distributed.is_initialized():
+        torch.distributed.destroy_process_group()
+
+
+def _dist_run_target(target: Callable, rank: int, num_procs: int, *args, **kwargs):
+    os.environ['WORLD_SIZE'] = str(num_procs)
+    os.environ['RANK'] = str(rank)
+    os.environ['LOCAL_RANK'] = str(rank)
+    os.environ['LOCAL_WORLD_SIZE'] = str(num_procs)
+
+    os.environ['MASTER_ADDR'] = '127.0.0.1'
+    os.environ['MASTER_PORT'] = str(29400)
+
+    target(*args, **kwargs)
+
+
+def with_distributed(num_procs: int,
+                     target: Callable,
+                     initial_timeout: int = 30,
+                     subsequent_timeout_per_device: int = 5):
+
+    def run_target(*args, **kwargs):
+        processes = []
+        ctx = mp.get_context('spawn')
+        for rank in range(num_procs):
+            _args = (target, rank, num_procs, *args)
+            p = ctx.Process(target=_dist_run_target, args=_args, kwargs=kwargs)
+            p.start()
+            processes += [p]
+
+        processes[0].join(initial_timeout)
+        for p in processes[1:]:
+            p.join(subsequent_timeout_per_device)
+
+        for rank, p in enumerate(processes):
+            if p.exitcode is None:
+                p.kill()
+                pytest.fail(f'Process {rank} did not exit', pytrace=False)
```

Comment:
Will we get stdout / stderr output from the sub processes?

Reply:
We do!